File: app/src/main/java/soup/movie/ui/main/settings/SettingsPresenter.java
package soup.movie.ui.main.settings;

import com.jakewharton.rxrelay2.BehaviorRelay;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import soup.movie.data.model.TheaterCode;
import soup.movie.settings.HomeTypeSetting;
import soup.movie.settings.TheaterSetting;
import soup.movie.ui.BasePresenter;
import soup.movie.ui.main.settings.SettingsViewState.DoneState;
import timber.log.Timber;

public class SettingsPresenter extends BasePresenter<SettingsContract.View>
        implements SettingsContract.Presenter {

    private final HomeTypeSetting homeTypeSetting;
    private final TheaterSetting theaterSetting;

    private final BehaviorRelay<Boolean> homeTypeSubject;

    @Inject
    SettingsPresenter(HomeTypeSetting homeTypeSetting,
                      TheaterSetting theaterSetting) {
        this.homeTypeSetting = homeTypeSetting;
        this.theaterSetting = theaterSetting;
        homeTypeSubject = BehaviorRelay.createDefault(homeTypeSetting.isVerticalType());
    }

    @Override
    protected void initObservable(CompositeDisposable subscriptions) {
        subscriptions.add(getViewStateObservable()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(view::render, Timber::e));
    }

    private Flowable<SettingsViewState> getViewStateObservable() {
        return Flowable.combineLatest(
                getHomeTypeObservable(),
                getFavoriteTheatersObservable(),
                DoneState::new);
    }

    private Flowable<Boolean> getHomeTypeObservable() {
        return homeTypeSubject.toFlowable(BackpressureStrategy.LATEST)
                .distinctUntilChanged();
    }

    private Flowable<List<TheaterCode>> getFavoriteTheatersObservable() {
        return Flowable.fromCallable(theaterSetting::getFavoriteTheaters);
    }

    @Override
    public void onVerticalHomeTypeClicked() {
        homeTypeSetting.setVerticalType();
        homeTypeSubject.accept(true);
    }

    @Override
    public void onHorizontalHomeTypeClicked() {
        homeTypeSetting.setHorizontalType();
        homeTypeSubject.accept(false);
    }
}

File: app/src/main/java/soup/movie/ui/main/settings/SettingsPresenter.kt
package soup.movie.ui.main.settings

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.internal.disposables.DisposableContainer
import soup.movie.data.MoopRepository
import soup.movie.data.helper.VersionHelper.currentVersion
import soup.movie.settings.impl.TheatersSetting
import soup.movie.ui.LegacyBasePresenter
import soup.movie.ui.main.settings.SettingsContract.Presenter
import soup.movie.ui.main.settings.SettingsContract.View
import soup.movie.ui.main.settings.SettingsViewState.TheaterListViewState
import soup.movie.ui.main.settings.SettingsViewState.VersionViewState

class SettingsPresenter(
    private val theatersSetting: TheatersSetting,
    private val repository: MoopRepository
) : LegacyBasePresenter<View>(), Presenter {

    override fun initObservable(disposable: DisposableContainer) {
        super.initObservable(disposable)
        disposable.add(repository.refreshVersion().subscribe())
        disposable.add(theatersSetting.asObservable()
            .map { TheaterListViewState(it) }
            .distinctUntilChanged()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { view?.render(it) })
        disposable.add(repository.getVersion()
            .startWith(currentVersion())
            .defaultIfEmpty(currentVersion())
            .onErrorReturnItem(currentVersion())
            .distinctUntilChanged()
            .map { VersionViewState(currentVersion(), it) }
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { view?.render(it) })
    }
}
